pkg/utils/vgs.go

```diff
@@ -0,0 +1,110 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+
+	crdv1alpha1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumegroupsnapshot/v1alpha1"
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// VolumeSnapshotParentGroupIndex is the name of the cache index hosting the relationship
+// between volume snapshots and their volume group snapshot owner
+const VolumeSnapshotParentGroupIndex = "ByVolumeGroupSnapshotMembership"
+
+// getVolumeGroupSnapshotParentObjectName returns the name of the parent group snapshot, if present.
+// The second return value is true when the parent object have been found, false otherwise.
+func getVolumeGroupSnapshotParentObjectName(snapshot *crdv1.VolumeSnapshot) string {
+	if snapshot == nil {
+		return ""
+	}
+
+	apiVersion := fmt.Sprintf(
+		"%s/%s",
+		crdv1alpha1.SchemeGroupVersion.Group,
+		crdv1alpha1.SchemeGroupVersion.Version,
+	)
+
+	for _, owner := range snapshot.ObjectMeta.OwnerReferences {
+		if owner.Kind == "VolumeGroupSnapshot" && owner.APIVersion == apiVersion {
+			return owner.Name
+		}
+	}
+
+	return ""
+}
+
+// IsVolumeGroupSnapshotMember returns true if the passed VolumeSnapshot object
+// is a member of a VolumeGroupSnapshot.
+func IsVolumeGroupSnapshotMember(snapshot *crdv1.VolumeSnapshot) bool {
+	parentName := getVolumeGroupSnapshotParentObjectName(snapshot)
+	return len(parentName) > 0
+}
+
+// VolumeSnapshotParentGroupKeyFunc maps a member snapshot to the name
+// of the parent VolumeGroupSnapshot
+func VolumeSnapshotParentGroupKeyFunc(snapshot *crdv1.VolumeSnapshot) string {
+	parentName := getVolumeGroupSnapshotParentObjectName(snapshot)
+	if len(parentName) == 0 {
+		return ""
+	}
+
+	return VolumeSnapshotParentGroupKeyFuncByComponents(types.NamespacedName{
+		Namespace: snapshot.Namespace,
+		Name:      parentName,
+	})
+}
+
+// VolumeSnapshotParentGroupKeyFuncByComponents computes the index key for a certain
+// name and namespace pair
+func VolumeSnapshotParentGroupKeyFuncByComponents(objectKey types.NamespacedName) string {
+	return fmt.Sprintf("%s^%s", objectKey.Namespace, objectKey.Name)
+}
+
+// NeedToAddVolumeGroupSnapshotOwnership checks if the passed snapshot is a member
+// of a volume group snapshot but the ownership is missing
+func NeedToAddVolumeGroupSnapshotOwnership(snapshot *crdv1.VolumeSnapshot) bool {
+	// Group snapshot ownership should be added only when
+	// the snapshot is marked with its group snapshot name
+	if snapshot == nil ||
+		snapshot.Status == nil ||
+		snapshot.Status.VolumeGroupSnapshotName == nil ||
+		len(*snapshot.Status.VolumeGroupSnapshotName) == 0 {
+		return false
+	}
+
+	parentObjectName := getVolumeGroupSnapshotParentObjectName(snapshot)
```

Comment:
Should we call getVolumeGroupSnapshotParentObjectName in the beginning of this function?  If parent object name is already set, we can just return false quickly. In dynamic provisioning case, it should already be set when the VS is first created.

If it is not set, then we check snapshot Status and decide whether to add it or not.

Reply:
Yes, looks good now.